### PR TITLE
fix(HMS-879): workers metrics

### PR DESCRIPTION
**Pull request by lzap**
This adds a new duration metric which captures duration of every job processed from the background redis queue. It contains "type" label which is the job name, e.g. `launch_instances_aws` for AWS EC2 launch job. Since this is the first metric we have in workers, the patch adds a new prometheus endpoint in the main function.

### CODE REVIEW

File: cmd/pbapi/main.go [hunk 1]
```diff
@@ -181,7 +181,7 @@ func main() {
 
 	if err := apiServer.ListenAndServe(); err != nil {
 		if !errors.Is(err, http.ErrServerClosed) {
-			log.Fatal().Err(err).Msg("Main service listen error")
+			log.Warn().Err(err).Msg("Main service listen error")
```
Comment (lzap) [inline cmd/pbapi/main.go:184]: I noticed that we call Fatal when there is a shutdown error. Fatal immediately exists the application will kill other goroutines. We should never use Fatal unless we want to exit immediately (e.g. in migrations or during startup).

**Comment by lzap** [comment]
Rebased to solve the import cycle problem.

File: internal/metrics/metrics.go [hunk 2]
```diff
@@ -3,8 +3,6 @@ package metrics
 import (
 	"time"
 
-	"github.com/RHEnVision/provisioning-backend/internal/kafka"
-	"github.com/RHEnVision/provisioning-backend/internal/models"
 	"github.com/RHEnVision/provisioning-backend/internal/version"
 	"github.com/prometheus/client_golang/prometheus"
```
Comment (lzap) [inline internal/metrics/metrics.go:7]: This was the culprit of import cycle problems. The metrics package will be in wide use througout the whole app and it is important to keep the imports at minimum. Therefore instead of `models.ProviderType` function argument, I refactored it to simply accept `string` because in the end, we convert the provider type to string anyway. This removes the need for import. And the same applies for everything, well, except `version` which is just a simple package with constants (no metrics there).

File: cmd/pbstatuser/main.go [hunk 3]
```diff
@@ -144,7 +144,7 @@ func checkSourceAvailabilityAWS(ctx context.Context) {
 
 	for s := range chAws {
 		logger.Trace().Msgf("Checking AWS source availability status %s", s.SourceApplicationID)
-		metrics.ObserveAvailabilityCheckReqsDuration(models.ProviderTypeAWS, func() error {
+		metrics.ObserveAvailabilityCheckReqsDuration(models.ProviderTypeAWS.String(), func() error {
```
Comment (adiabramovitch) [inline cmd/pbstatuser/main.go:147]: Why did you add `.String()`?
it is not needed, it is present [here](https://github.com/RHEnVision/provisioning-backend/blob/bad43f81bf343b9f3898ab69ded82b6198804bd2/internal/metrics/metrics.go#L55)

`provisioning_source_availability_check_request_duration_ms_bucket{component="statuser",error="true",service="provisioning",type="aws",le="5000"} 2`

File: internal/metrics/metrics.go [hunk 4]
```diff
@@ -48,25 +46,44 @@ var AvailabilityCheckReqsDuration = prometheus.NewHistogramVec(
...
+		Buckets:     []float64{5, 10, 50, 100, 250, 500, 1000, 2500, 5000, 10000},
+	},
+	[]string{"type"},
+)
+
+func ObserveAvailabilityCheckReqsDuration(provider string, observedFunc func() error) {
 	errString := "false"
 	start := time.Now()
 	defer func() {
-		AvailabilityCheckReqsDuration.WithLabelValues(provider.String(), errString).Observe(float64(time.Since(start).Nanoseconds()) / 1000000)
```
Comment (adiabramovitch) [inline internal/metrics/metrics.go:55]: Why?
I think it is cleaner to put it here

File: internal/metrics/metrics.go [hunk 5]
```diff
@@ -88,3 +105,7 @@ func RegisterStatuserMetrics() {
 func RegisterApiMetrics() {
 	prometheus.MustRegister(JobQueueSize)
 }
+
+func RegisterWorkerMetrics() {
+	prometheus.MustRegister(JobQueueSize)
```
Comment (adiabramovitch) [inline internal/metrics/metrics.go:110]: it is supposed to be `BackgroundJobDuration` isnt it?

File: pkg/worker/redis.go [hunk 6]
```diff
@@ -205,7 +206,9 @@ func (w *RedisWorker) processJob(ctx context.Context, job *Job) {
 	logger.Info().Msg("Dequeued job from Redis")
 	ctx = contextLogger(ctx, job)
 	if h, ok := w.handlers[job.Type]; ok {
-		h(ctx, job)
+		metrics.ObserveBackgroundJobDuration(job.Type.String(), func() {
```
Comment (ezr-ondrej) [inline pkg/worker/redis.go:209]: ouch. we are using internal package function in `pkg`. Can't we do it in `job/common.go` somewhere?
Or just abandone the idea of it being under `pkg`.
Not a blocker, more of a food for thought, given we already import other internal packages :shrug:

File: internal/metrics/metrics.go (hunk 5 above)
Comment (lzap) [reply internal/metrics/metrics.go:110]: Damn good eyes! I tested only that the metrics appear in general, but it actually DOES NOT WORK without this.

File: pkg/worker/redis.go (hunk 6 above)
Comment (lzap) [reply pkg/worker/redis.go:209]: Yeah this is not the first time. Tho I think it is sort of okay I guess as long as you so not return an "internal" type. Unless you prove otherwise, I think this is okay.

I think somebody from the content team already asked about our Redis queue and I gave them link. Maybe this would get used by someone else?!?

We can always move this into the internal package later.

**Comment by lzap** [comment]
Changed one line which caused this patch to, ehm, not work. :-) Copy and paste...

File: internal/metrics/metrics.go (hunk 5 above)
Comment (adiabramovitch) [reply internal/metrics/metrics.go:110]: What about registering `JobQueueInFlight`?

**Comment by lzap** [comment]
Also fixed queue size which was not working (it was in the wrong initialization function).

File: internal/metrics/metrics.go [hunk 7]
```diff
@@ -86,5 +103,10 @@ func RegisterStatuserMetrics() {
 }
 
 func RegisterApiMetrics() {
+	// no metrics
+}
+
+func RegisterWorkerMetrics() {
 	prometheus.MustRegister(JobQueueSize)
+	prometheus.MustRegister(BackgroundJobDuration)
```
Comment (adiabramovitch) [inline internal/metrics/metrics.go:111]: Just for consistency
```suggestion
	prometheus.MustRegister(JobQueueSize, BackgroundJobDuration)
```

Comment (ezr-ondrej) [reply internal/metrics/metrics.go:111]: @adiabramovitch could you do a follow-up for it? If it's not a blocker, I'd like to get this in, so we start capturing the metrics and Lukas will not get to it of a week now :)

**Review by adiabramovitch — APPROVED** [review]

File: internal/metrics/metrics.go (hunk 7 above)
Comment (ezr-ondrej) [reply internal/metrics/metrics.go:111]: Addressed in #443
